agb/src/display/affine.rs

```diff
@@ -0,0 +1,223 @@
+    /// Generates the matrix that represents a translation by the position
+    #[must_use]
+    pub fn from_position(position: Vector2D<Num<i32, 8>>) -> Self {
+        AffineMatrix {
+            a: 1.into(),
+            b: 0.into(),
+            c: 0.into(),
+            d: 1.into(),
+            x: position.x,
+            y: position.y,
+        }
+    }
+
+    #[must_use]
+    /// The position fields of the matrix
+    pub fn position(&self) -> Vector2D<Num<i32, 8>> {
+        (self.x, self.y).into()
+    }
+
+    /// Attempts to convert the matrix to one which can be used in affine
+    /// backgrounds.
+    pub fn try_to_background(&self) -> Result<AffineMatrixBackground, OverflowError> {
+        Ok(AffineMatrixBackground {
+            a: self.a.to_raw().try_into().map_err(|_| OverflowError(()))?,
+            b: self.a.to_raw().try_into().map_err(|_| OverflowError(()))?,
+            c: self.a.to_raw().try_into().map_err(|_| OverflowError(()))?,
+            d: self.a.to_raw().try_into().map_err(|_| OverflowError(()))?,
+            x: self.a.to_raw(),
+            y: self.a.to_raw(),
+        })
+    }
+
+    #[must_use]
+    /// Converts the matrix to one which can be used in affine backgrounds
+    /// wrapping any value which is too large to be represented there.
+    pub fn to_background_wrapping(&self) -> AffineMatrixBackground {
+        AffineMatrixBackground {
+            a: self.a.to_raw() as i16,
+            b: self.a.to_raw() as i16,
+            c: self.a.to_raw() as i16,
+            d: self.a.to_raw() as i16,
+            x: self.a.to_raw(),
+            y: self.a.to_raw(),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[repr(C, packed(4))]
+/// An affine matrix that can be used in affine backgrounds
+pub struct AffineMatrixBackground {
+    // Internally these can be thought of as Num<i16, 8>
+    a: i16,
+    b: i16,
+    c: i16,
+    d: i16,
+    // These are Num<i32, 8>
+    x: i32,
+    y: i32,
+}
+
+impl TryFrom<AffineMatrix> for AffineMatrixBackground {
+    type Error = OverflowError;
+
+    fn try_from(value: AffineMatrix) -> Result<Self, Self::Error> {
+        value.try_to_background()
+    }
+}
+
+impl AffineMatrixBackground {
+    #[must_use]
+    /// Converts to the affine matrix that is usable in performing efficient
+    /// calculations.
+    pub fn to_affine_matrix(&self) -> AffineMatrix {
+        AffineMatrix {
+            a: Num::from_raw(self.a.into()),
+            b: Num::from_raw(self.b.into()),
+            c: Num::from_raw(self.c.into()),
+            d: Num::from_raw(self.d.into()),
+            x: Num::from_raw(self.x),
+            y: Num::from_raw(self.y),
+        }
+    }
+}
+
+impl From<AffineMatrixBackground> for AffineMatrix {
+    fn from(mat: AffineMatrixBackground) -> Self {
+        mat.to_affine_matrix()
+    }
+}
+
+impl Default for AffineMatrix {
+    fn default() -> Self {
+        AffineMatrix::identity()
+    }
+}
+
+impl Mul for AffineMatrix {
```

Comment:
also worth implementing addition?

Reply:
what does addition mean...

Reply:
fair. I guess it is a faster way to add translations but that probably won't come up very often...

Reply:
Translation has identity for the rot/scale part, so that would not work too great

Reply:
ah yes. In which case it would be a bit pointless :P